Alphabetize imports within groups

stropho opened the pull request:
I'm not the real author of these commits. I just rebased the existing https://github.com/benmosher/eslint-plugin-import/pull/1105 . I'm looking forward to use this new feature.

Closes #1105. Fixes #1406.

stropho commented:
Can anybody help? Is there any way to test incorrect configuration passed to the eslint rule? It seems like such errors are handled already by the `RuleTester` and not the rule itself, so I don't see i simple way to test it at the moment.

Or should we just delete the checks to make coveralls happy ?

ljharb commented:
Please don’t open duplicate PRs; now have to maintain both in sync.

Instead, spot a link to your commits on the original PR, and a maintainer will update it.

stropho commented:
ok, got it

VincentLanglet commented:
Is this ready to merge ? This feature will be great.

stropho commented:
I believe it is ready. I created a fix based on the suggestion in https://github.com/benmosher/eslint-plugin-import/pull/1105 . And now, probably some of the maintainers have to say "it's ok" and merge it.

syashenkov commented:
I think example still needs correction:
```js
import assert from 'assert';
import buffer from 'buffer';
import foo from 'foo';
import bar from 'bar';
import Baz from 'Baz';
```
It's unclear if imports should be sorted by module name or imported names. Also, it's uncertain what behaviour is expected using star imports (import * as) and symbol imports (import { A, B })

It would be best to make example with some real life imports like
```js
import * as classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { compose } from 'recompose';
```

syashenkov commented:
Or a more clear example like

Pass:
```js
import c from 'a';
import b from 'b';
import a from 'c';
```
Fail:
```js
import a from 'c';
import b from 'b';
import c from 'a';
```

syashenkov commented:
@ljharb Could you please look at this PR?

stropho commented:
Fixed the example. 
Rebased on master branch. 
Hope everybody is happy now.

Travis CI failed but it has the same errors as master branch did https://travis-ci.org/benmosher/eslint-plugin-import/builds/553254212 . So I don't think there is anything I can do about it right now.

deb0ch commented:
The tests don't seem to be failing on `master` anymore, is this mergeable soon before there are conflicts ?

stropho commented:
@deb0ch should be mergable; thanks for the reminder, just rebased :champagne: :tada: :champagne: :tada:

luczsoma commented:
@ljharb Could you look into merging this PR?

It is a long-wanted feature, and also would come with a few benefits, since it:
- Fixes #389 
- Closes #629 
- Closes #1105 

Thank you.

syabro commented on src/rules/order.js:
```diff
@@ -240,17 +241,61 @@ function makeOutOfOrderReport(context, imported) {
+    groupedByRanks[groupRank].sort(function(importA, importB) {
+      return ignoreCase
+        ? importA.localeCompare(importB)
+        : (importA < importB ? -1 : (importA === importB ? 0 : 1))
```
Scary. Could it be replaced with `if... return`?

luczsoma replied:
Btw it is also incorrect I think. The default `sensitvity` option is `variant` in `localeCompare` for the default `sort`ing usage (applies in the above snippet), which [does not ignore casing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare#Parameters).

I don't want to meddle in this PR, but I suggest the following:
```
groupedByRanks[groupRank].sort(function(importA, importB) {
  return importA.localeCompare(importB, undefined, { sensitivity: ignoreCase ? 'accent' : 'case' })
})
```

luczsoma commented:
@stropho, since I have no write access to this, in order to get this PR finally merged before any further conflicts emerge, I would like to ask you to

- either commit the following two minor changes, please,
- or invite me as a contributor into `stropho/eslint-plugin-import-1` so I am able to commit them.

1. In `src/rules/order.js`, change this:

```
groupedByRanks[groupRank].sort(function(importA, importB) {
  return ignoreCase
    ? importA.localeCompare(importB)
    : (importA < importB ? -1 : (importA === importB ? 0 : 1))
})
```

to this:

```
groupedByRanks[groupRank].sort(function(importA, importB) {
  return importA.localeCompare(importB, undefined, { 
    sensitivity: ignoreCase ? 'accent' : 'case' 
  })
})
```

2. Add the following to `CHANGELOG.md`, into the Unreleased section:

```
- [`order`]: Add support for alphabetical sorting within import groups ([#1360], thanks [@duncanbeevers] & [@stropho])
```

Thank you!

stropho commented:
@luczsoma  I haven't really investigated the code much, just wanted to help to merge it.... anyway, I just tried to fix it myself since you've told me exactly what to do :-D About 4 tests fail now. That is ok but it seems to be nearly impossible for me to find which tests are failing. 
I added you as collaborator... Feel free to add changes. I could find some spare time at the end of the week.

luczsoma commented:
Thank you @stropho!

As I tried to extend the functionality with configurable numeric sorting (1 < 2 < 10 vs 1 < 10 < 2) and configurable casing options (a-z < A-Z vs a-z > A-Z), I found out that `localeCompare` doesn't really work as common sense expects, see https://stackoverflow.com/questions/57671159. For the sake of simplicity, I removed `localeCompare`.

I also fixed @syabro's concerns.

luczsoma commented:
I think this should be mergeable now.

syabro commented:
> As I tried to extend the functionality with configurable numeric sorting (1 < 2 < 10 vs 1 < 10 < 2) and configurable casing options (a-z < A-Z vs a-z > A-Z)

I bet it could be done later :) We badly need a basic sort

luczsoma commented:
> > As I tried to extend the functionality with configurable numeric sorting (1 < 2 < 10 vs 1 < 10 < 2) and configurable casing options (a-z < A-Z vs a-z > A-Z)
> 
> I bet it could be done later :) We badly need a basic sort

Planning to do it some time later. :)

stropho commented:
thx @luczsoma  - I'm not a fan of rewriting code to make it bigger but as long as it is more readable and everybody's happy about it.. why not...

What got my attention though is that the `ignoreCase` probably does not work as it was intended. 
My understanding when `ignoreCase=true`: it doesn't matter if `a` is before or after `A`
My understanding of how it works now: sorting treats `a` that it equals `A` but the tested code needs to be equal to the result of sorting :-/
So I changed the tests a bit to actually test the `ignoreCase` option and had to comment 2 of them because it seems they are not fixable with the current implementation. So what now ?  

1. easier choice - remove `ignoreCase` option. Why would anyone create files with capital or non-ascii letters ? I understand that this is strongly opinionated. But if anyone will ever want such feature, it could be added later. As mentioned before:
> We badly need a basic sort
2. fix the `ignoreCase` case. This does not quite fit my schedule. But maybe someone else can try ?

Let's vote!!! 
:+1: for option 1.
 :rocket:  for option 2.

stropho commented:
alright, `ignoreCase` option is removed. Less options --> smaller chance something is not working :wink:

stropho commented:
@ljharb any news? any suggestions to get it merged ?

aunpongsit commented:
any updates on this? really want it to be merged.

ljharb commented on src/rules/order.js:
```diff
@@ -240,6 +240,50 @@ function makeOutOfOrderReport(context, imported) {
+}
+function mutateRanksToAlphabetize(imported, order) {
+  const groupedByRanks = imported.reduce(function(acc, importedItem) {
+    acc[importedItem.rank] = acc[importedItem.rank] || []
```
```suggestion
    if (!Array.isArray(acc[importedItem.rank]) { acc[importedItem.rank] = [] }
```

dietergeerts replied:
> 

I was just looking into this PR because I can use it, though I wonder for what reason you find the suggested change better? For my personally, what's there is more readable.

ljharb replied:
avoiding unconditional assignment is more correct, which is more important than what's readable.

ljharb commented on src/rules/order.js:
```diff
@@ -384,6 +435,15 @@ module.exports = {
             ],
           },
+          alphabetize: {
+            type: 'object',
```
```suggestion
            type: 'object',
            additionalProperties: false,
```

luczsoma replied:
What is the change request here? To move the `additionalProperties: false` directly under `type: 'object'`, instead of its current position at the bottom of the schema (under `properties`)?

ljharb commented on tests/src/rules/order.js:
```diff
@@ -335,10 +335,10 @@ ruleTester.run('order', rule, {
     test({
       code: `
-        import foo
+        import a
```
why does this need to be changed? can this test be reverted?

luczsoma replied:
I think the intention of this commit (which is not my own) was to show how the alphabetical sorting in a clearer way, with minimal examples. I would not revert that.

luczsoma replied:
If your comment is concerned about only this particular test case (which tests multiline, not ordering), I will revert that part to its original state.

ljharb replied:
Yes, my comment is because for it not to be a breaking change, no existing test cases should be altered.

ljharb commented:
(additionally, please rebase this on latest master)

luczsoma commented:
Requested changes are done, rebased onto master.

luczsoma commented:
@ljharb Is there anything else we can do to get this merged? I cannot dismiss the change request (although everything requested is done), only you (or the author of this PR, @stropho) can.

nmorel commented:
I tried the PR and it breaks the `newlines-between` rules.  
With `always`, I have an empty line between all my imports and not just groups.  
From what I see, `newlines-between` uses rank to determine empty lines to add and alphabetize modifies the ranks to order.

nmorel commented:
Putting the newlines test first seems to fix it :   
```javascript
if (newlinesBetweenImports !== 'ignore') {
  makeNewlinesBetweenReport(context, imported, newlinesBetweenImports)
}

if (alphabetize.order !== 'ignore') {
  mutateRanksToAlphabetize(imported, alphabetize.order)
}

makeOutOfOrderReport(context, imported)
```

cameronmurphy commented:
Any news on this? Very keen for this functionality!

kondei commented:
I want this

iippis commented:
You guys are so close to the finish line, don't quit now! I'm eagerly waiting for this feature.
